### ISSUE TITLE
Lighting Update Implementation

### ISSUE THREAD

**Issue by codewarrior0** [mcedit/mcedit2#117]
I am trying to come up with a new implementation for MCEdit's lighting update that is fast enough to be processed immediately. I'll begin with a short summary of the in-game lighting algorithm, and follow by describing my past implementations.

# Algorithm

The behavior of light is simple enough to intuit: Blocks such as torches "glow" with their own light level. Block cells adjacent to a glowing block have a light level one less than that, and cells adjacent to those have a light one less than that, and so on until the light level drops to zero. Solid blocks such as stone are completely opaque and drop the light level to zero immediately. Others such as ice and water are partially opaque and drop the light by more than one level.

Lighting must be updated whenever a block is changed, and we can consider only a few general cases:

The simplest case is when a block is changed into a light source, or when a light source is changed to a brighter source. First, set that cell's light value to the light source value. Then, compare that cell with the six cells adjacent. For each adjacent cell, subtract its opacity value (which is always at least 1) from the current cell's light, and then set the adjacent cell's light to the new value. If the adjacent cell's light value has changed, repeat the process with the adjacent cell as the new current cell. 

In pseudocode:

(For brevity, only the x coordinate is used and only the x-1 neighbor is checked. In real code, we need to check x-1, x+1, y-1, y+1, z-1, z+1)

```
def blockChanged(x):
    light = blockBrightness(x)
    setLight(x, light)
    spreadLight(x)

def spreadLight(x):
    light = getLight(x)
    if light <= 0:
        return

    # Repeat for all six adjacent cells

    adjacentLight = getLight(x-1)
    adjacentOpacity = blockOpacity(x-1)

    # If the adjacent cell already has the "correct" light value, stop.
    if light - adjacentOpacity > adjacentLight:
        setLight(x-1, light - adjacentOpacity)
        spreadLight(x-1)

    # End repeat
```

A variant of the first case is when a block is changed to another block with a lesser opacity value. For example, replacing ice with air. In this case, instead of replacing this cell's light value with the new block's light value (as it doesn't have one), we will recompute this cell's light value from its brightest neighbor. The cell's light is set to its brightest neighbor's light minus its own opacity, and then the light is updated as before:

```
def blockChanged(x):
    previousLight = getLight(x)
    light = blockBrightness(x)
    setLight(x, light)
    opacity = blockOpacity(x)

    # Repeat for all six adjacent cells

    adjacentLight = getLight(x-1)
    if adjacentLight - opacity > getLight(x):
        setLight(x, adjacentLight - opacity)

    # End repeat

    if previousLight != getLight(x):
        spreadLight(x)
```

The last case is when a block is changed to a dimmer light source. This also includes the cases where a block is changed to one with a greater opacity value. Both cases have the same result: the cell's light value is decreased. To handle this case correctly, we must first find every block that was previously illuminated by the cell's old light value and reset its light using its neighbor's light as in the variant of the first case. Then, each of those cells that was reset to a nonzero light will cast its light into adjacent cells using spreadLight().

Since all of the cells touched by the previous light source still have their old light values, we must do this in several passes. First, find all cells touched by the previous light source, set their light values to zero, and store their positions. Second, set the light value of every cell found to the light value of its brightest neighbor minus its own opacity. Third, for every cell whose light changed in the second step, cast its light using spreadLight().

This search will find the cells adjacent to the changed block and has the effect of spreading the changed block's new light value.

This is the complete algorithm:
```
def blockChanged(x):
    previousLight = getLight(x)
    light = blockBrightness(x)
    setLight(x, light)

    drawLight(x)

    if previousLight > getLight(x):
        spreadLight(x)

    if previousLight < getLight(x):
        fadeLight(x, previousLight)


def drawLight(x):
    opacity = blockOpacity(x)

    # Repeat for all six adjacent cells

    adjacentLight = getLight(x-1)
    if adjacentLight - opacity > getLight(x):
        setLight(x, adjacentLight - opacity)

    # End repeat

def spreadLight(x):
    light = getLight(x)
    if light <= 0:
        return

    # Repeat for all six adjacent cells

    adjacentLight = getLight(x-1)
    adjacentOpacity = blockOpacity(x-1)

    # If the adjacent cell already has the "correct" light value, stop.
    if light - adjacentOpacity > adjacentLight:
        setLight(x-1, light - adjacentOpacity)
        spreadLight(x-1)

    # End Repeat


def fadeLight(x, previousLight):
    fadedCells = findFadedCells(x, previousLight)
    for x in fadedCells:
        drawLight(x)
    for x in fadedCells:
        spreadLight(x)


def findFadedCells(x, previousLight)
    fadedCells = []
    light = getLight(x)

    # Repeat for all six adjacent cells

    adjacentLight = getLight(x-1)
    adjacentOpacity = blockOpacity(x-1)
    if previousLight - adjacentOpacity == adjacentLight:
        fadedCells.append((x-1))
        setLight(x-1, 0)
        fadedCells.extend(findFadedCells(x-1, adjacentLight))

    # End Repeat

    return fadedCells
```

# MCEdit 1.0 Implementation

MCEdit 1.0 used a completely different approach to updating the light values after blocks are changed. Instead of dealing with individual cells, the implementation updates light values across the entire chunk. First the entire chunk is reset to the brightness values of each block. Numpy's array slicing is used to get the entire chunk's light without its right edge, as `left`. Then, the entire chunk's opacity values without its left edge, as `right`. Then, `left - right` gives the new light values for all the blocks in `right`. This is repeated for all six directions, and for 15 iterations. Temporary arrays are used to hold the left and right edges of the adjacent chunks and detect whether the changed light has spilled into those chunks so they can be updated likewise.

When a single block is changed, its chunk is marked for lighting updates, and since a light change has effects up to 15 blocks away in all directions, the eight adjacent chunks are also marked. 

This approach has the obvious drawback of updating far more cells than is strictly necessary. However, it also has the benefit of not dealing with block coordinates at all. By doing this, it also sidesteps the inefficiency of checking a single cell for updates from multiple directions. (Imagine the light from one block spreading up and then left, and that same light also spreading left and then up to touch the other block twice.) 

I haven't bothered to estimate it, but for large numbers of simultaneous block changes, there has to be a break-even point at which the chunk-by-chunk method is more time-efficient (and space-efficient!) than the block-by-block method. To give you an idea of where it is, placing a single glowstone block in the middle of a large enough dark cavern will change the light values of ~15,000 cells; a Minecraft Alpha chunk (at the time the algorithm was written) has 32,768 cells.

# Array-based setBlocks implementation

MCEdit 2.0 has a partial, and very broken implementation that builds upon the new `setBlocks` function, which allows any number of blocks (and light values, biomes, etc) to be changed in a single call. It is roughly equivalent to the first block-by-block except that the individual x, y, z coordinates are arrays of coordinates instead. Even though it is incomplete - it only implements the first case of "block light increased" - it already reveals some shortcomings. The biggest one being that there is no elimination of duplicate coordinates. 

(going to bed - will finish later)

**Comment by maruohon** [comment]
It's a bit hard to grasp it all after first reading through it and know if I'm completely in the dark with my ideas, but here goes.

So to me it seems that increasing the light value, ie. spreading new light is much easier and straight forward? And the removing or dimming of light sources is the more complicated case?

Bases on that my first thought (which I'm not sure if it is already what is happening) is something like this:
When a light source ir removed (or dimmed), first decrease the light value of all the blocks in a diamond-ish shape around it (or just a box?), by the amount that would possibly have been received from that light source if completely non obstructed (ie. Air; so basically decrease the light value by the removed source's light value minus the block distance from it, or even straight to zero?). Then re-spread the light (only needs to be done "inwards" towards the area?) from all the light sources that are in a 15-ish block range from the edges of that dimmed diamond-shaped area (or box) that was affected.

Not sure if this makes any sense or would be cheaper than the existing algorithms. My thinking was that it might be less complex than trying to find out the blocks that actually got affected, just reset everything to zero in the range possibly affected and re-spread from the surroundings and the affectedarea's remaining light sources.

**Comment by CAD97** [comment]
I've thought about light propagation before, and since you want ideas no matter how far-fetched, I'll share what I've thought about that seems relatively OK. You might be able to use something.

(This will be pseudocode in a C#/Java like style since that's what I'm used to.)

I always saw MineCraft light propagation as a recursive problem (which is strange, given my opinion on recursion -- that it's rarely applicable). In essence, each block would be responsible for notifying it's neighbors when its light propagation changed. Thus, when a block's light value changed by the player:

```
void setLightLevel(int value)
{
    if (value == this.lightLevel) return;
    //else
    bool increased = value > this.lightLevel;
    this.lightLevel == value;
    foreach (BlockPos pos in this.pos.neighbors) {
        pos.block.setLightLevel(this.pos, this.lightLevel, increased);
    }
}
```

Thus, the actual block's change is quick to process and notifies the neighboring blocks that they need to update their light via setLightLevel(BlockPos, int, bool). This method is in charge of handling propagating light updates away from the source of the update and terminating them if no change occurs. Thus:

```
void setLightLevel(BlockPos source, int sourceValue, bool increasing)
{
    if (increasing && sourceValue <= this.lightLevel) return;
    if (!increasing && sourceValue >= this.lightLevel) return;
    //else
    this.lightLevel == sourceValue - opacity;
    foreach (BlockPos pos in this.pos.neighbors) {
        if (pos != source) {
            pos.block.setLightLevel(this.pos, this.lightLevel, increased);
        }
    }
}
```

Disclaimer: I've not done any testing on this algorithm, just speculation, and may be just rehashing things you've already considered. I just wanted to share my $0.02. In addition, I have no idea how this stands up to the large use case that MCEdit is concerned with, as when I came up with this I was only thinking about handling in-game cases (before /fill).

------------------------------------------------------------------------------------------------------------------------

From the more theoretical side, I think I agree that the light increase algorithm is by nature 'cheaper' then the light decrease. In the case of losing light sources it _might_ be more efficient to just black-out 15 blocks away from any modified blocks and just propagate the light back in.

Thinking shortly about it, it's worth considering an algorithm similar to this:
* Black out changed blocks + 15 blocks away (unless light source)
* Propagate light away from light sources within blacked out area
* Propagate light in from the edges of blacked out area

Lastly, I'm ignoring sunlight because that's done on the fly by the engine, right?

**Comment by codewarrior0** [comment]
@maruohon:

That's a good point, but I don't see it as much of an improvement. You're asking if you can skip the opacity check when searching for blocks to dim, and finding all of the cells that the light would have touched if none of the blocks were opaque. In exchange for skipping the opacity check, you're now calling drawLight and spreadLight on a whole bunch of cells that aren't actually touched by the changed block's previous opacity.

However, you did remind me of an important flaw in my findFadedCells function. I was setting all of the cells to zero, when what I should do is set the cells to the block's brightness value instead, so it can re-propagate those light sources (for instance, if that light source was washed out by a brighter light source, which is now being removed). I'll edit that function appropriately.

**Comment by codewarrior0** [comment]
@CAD97

Thanks for the input. It's simple enough to mentally revise your example with the condition that there isn't actually a `Block` class where every section has 4,096 instances of `Block`. MCEdit only operates on the numeric values in the `Blocks` and `Light` arrays. I'll assume that your `this.lightLevel = foo` is a no-op for blocks that emit their own light, so as not to make the same mistake I did just now.

I agree, it seems like it's possible to dim the block's light in a single pass, but your dimming approach will halt when it reaches the edge of the new light value's "light diamond" and not go on to draw in light from the outer edges of the old light value's diamond, nor update any values outside of the new light diamond.

This is not necessarily a recursive problem. It can be considered as iterations of a function whose input is a list of coordinates where updates will be applied (plus the world state), and whose output is a list of coordinates where updates should be applied next (plus the world state). The MCEdit 2.0 implementation is roughly like this, calling a function repeatedly until the coordinate list it returns is empty.

Sunlight is handled by a simple pre-processing step that checks if the column's height changed and calling spreadLight or fadeLight as needed. Essentially, cells above the column's height are treated as light-emitting cells.

**Comment by codewarrior0** [comment]
@CAD97 to your edit:

You seem to be suggesting the same thing @maruohon did, skipping the opacity check in exchange for updating several cells that do not even need it. Thinking of a common case of a light source placed against a wall indoors, I'd expect that skipping the opacity check would update almost twice as many cells as is needed, if not more.

**Comment by codewarrior0** [comment]
I've implemented the base algorithm in pure python to use as a point of reference. I had to fix a mistake in `blockChanged` where the `<` and `>` were swapped, and changed `findFadedCells` to use a queue for the cells to check and a set for the cells already found, removing the recursive call, as it was spending far too much time catenating lists with duplicate entries. These changes are reflected in the opening post.

**Comment by codewarrior0** [comment]
I've implemented the base algorithm in Cython. It uses a `RelightCtx` class to wrap the WorldEditor and provide fast, cached, C++-level access to chunk section arrays while still holding references to the WorldEditor's chunks to keep them alive. It also uses STL containers in place of Python collection objects wherever possible.

It is at least 50 times faster than the pure python algorithm. I found that updateLights implicitly assumes that the cell's opacity has changed (as it doesn't have the cell's previous block ID), so I modified `copyBlocks()` with a couple of strategies for calling updateLights only on the cells whose opacity or brightness changed. The results and some further ponderings are in [`time_relight_manmade.py`](https://github.com/mcedit/mcedit2/blob/master/src/mceditlib/bench/time_relight_manmade.py)

In summary, updating lights across the entire copied area is faster than updating lights for each section after copying that section, because later copies will redo some of the work done by earlier light updates. Thus, a strategy for only relighting copied chunks after all adjacent chunks are copied will be needed, along with a strategy for managing the memory used for the block coordinates that need updates.

**Comment by codewarrior0** [comment]
Build 649 has a more-or-less complete lighting implementation. Light will update immediately after each edit, rather than all at once when the world is saved.

It is not as fast as I want it to be, though. I can tell that there is a lot of duplicate and wasted work done while updating the skylight after the height of a column changes.
